### PR TITLE
Replace Python core with pure-Vimscript core ("vimcore")

### LINKED ISSUE #57: escape pyscript_path
After upgrading, I now see an error:

```
E172: Only one file name allowed: pyfile /Users/josephholsten/Dropbox (Personal)/Mackup/.config/nvim/plugged/editorconfig-vim/plugin/editorconfig.py
```

I'm guessing this error is because the absolute path to my vim plugins includes whitespace. 

I'm thinking we should either escape [where pyscript_path is set `plugin/editorconfig.vim:39`](https://github.com/editorconfig/editorconfig-vim/commit/c2b7a104b826b7ff9283d32cb95a039ddccde79b#diff-b96ea8b60d0436a02a99512c2e5942fdR39) or [where it's passed to pyscript `plugin/editorconfig.vim:250`](https://github.com/editorconfig/editorconfig-vim/commit/c2b7a104b826b7ff9283d32cb95a039ddccde79b#diff-b96ea8b60d0436a02a99512c2e5942fdR250).

### LINKED ISSUE #101: Get errors when there is no +python feature.
This error should be catched, can use warn message instead. I will try to send a PR to fix it.

qq 20171227111822

### LINKED ISSUE #116: Wrong `EditorConfig_python_files_dir` when open file via command-line on Windows.
Call example:
```
C:\> vim "[FILEPATH]"
```

Error code:
```
Faild to execute " "python" "C:\/main.py" "[FILEPATH]"". Exit code: 2
Message:
['(null): can''t open file ''C:\/main.py'': [Errno 2] No such files or directory']
```

### PR DESCRIPTION
**Pull request by cxw42**
Here it is - let me know what you think!  Passes all tests except Windows UTF-8, which @ppalaga says might not be possible (https://github.com/editorconfig/editorconfig-core-c/pull/31#issue-154810185).

Closes https://github.com/editorconfig/editorconfig/issues/383 .
Closes #48 (replaces it).

Also closes the open Python issues:
Closes #57.
Closes #101.
Closes #116.

May help with #108.
May affect #111.

Part of the discussion in https://github.com/vim/vim/issues/2286 .

### CODE REVIEW

**Comment by cxw42** [comment]
The CI failure appears to be related to rake/bundler/vimrunner versions.  Would someone with some Ruby expertise be able to give me some suggestions?

**Comment by xuhdev** [comment]
I guess you have generated the lock file in a new version of bundler. The easiest way to resolve this is probably use bundler less than 2.0 to generate Gemfile.lock on your system.

**Comment by cxw42** [comment]
@xuhdev Thanks - that worked!  I was able to run the tests on Cygwin from `editorconfig-vim/` with

    export BUNDLE_GEMFILE=tests/Gemfile
    bundle _1.15.4_ exec rspec tests/spec/editorconfig_spec.rb

Please let me know if you would like any changes, or if you have any concerns.

**Comment by cxw42** [comment]
@xuhdev Any thoughts?  I am wondering if we can merge this one.  After that, I would like to work with @rakus to see if we can merge some of his plugin changes into this official plugin.  I think removing the Python should come before any Vimscript enhancements.  What say you?

**Comment by xuhdev** [comment]
I remember you said something about number range (`{-3...100}`) and this core depends on that. I wanted to start a voting for that I couldn't find that thread. Could you remind me where it is?

**Comment by cxw42** [comment]
@xuhdev 

> CI

The CI (Travis) is unchanged in this PR.  The core is copied from https://github.com/cxw42/editorconfig-core-vimscript , which has its own CI (Appveyor) set up.

> number range

The number range is editorconfig/editorconfig#371 .  Thank you for following up on it!

I changed this PR so it does not depend on that issue.  It has the current behaviour (e.g., `{-10..10}` rejects `0`).  Therefore, I think this can go ahead even before the vote on number ranges.

**Comment by xuhdev** [comment]
I understand you haven't changed anything about CI. What I meant is that, if you add a core library to the repo, we need CI to test that core library for guaranteeing its continuously correct functioning...

**Comment by cxw42** [comment]
@xuhdev Sorry for misunderstanding!  I have not but am happy to.  However, the vimscript core requires a number of support files in order to be testable by ctest (e.g., bash scripts and bat files).  May I add a separate repo submodule for those so plugin users don't get files they don't need?

**Comment by xuhdev** [comment]
I personally feel it's OK to add editorconfig/editorconfig-core-test as a submodule, but all other scripts IMO are integrated part of this subproject. Users in general get testing scripts, same for most other cases if they are supposed to get things from source. In addition, test scripts are unlikely to use a lot of spaces anyway.

**Comment by cxw42** [comment]
@xuhdev OK - I'll update this repo.  I may ask to revisit this point later if Bram decides he's willing to bundle this plugin with Vim.

**Comment by cxw42** [comment]
@xuhdev As of https://github.com/editorconfig/editorconfig-vim/pull/119/commits/e2bd6b86bacfefddbfb1f104f665824c9d4ce1ed , Travis tests both the core and the plugin.  I also have Appveyor testing the core so we have a Windows test case.  I cannot yet test the plugin on Windows since vimrunner doesn't support Windows (AndrewRadev/vimrunner#32).

File: README.md [hunk 1]
```diff
@@ -35,6 +28,10 @@ OR
 
   Then remember to call `:PluginInstall`.
 
+### No editorconfig core required!
```
Comment (xuhdev) [inline README.md]: Perhaps "No external editorconfig core library is required"?

File: autoload/editorconfig_core.vim [hunk 2]
```diff
@@ -0,0 +1,138 @@
+" autoload/editorconfig_core.vim: top-level functions for
+" editorconfig-core-vimscript.
+" Copyright (c) 2018 Chris White.  All rights reserved.
```
Comment (xuhdev) [inline autoload/editorconfig_core.vim]: Can you add a license header block for every file? You can refer to other files in the plugin for how to do it. Also, should the year be 2019?

File: autoload/editorconfig_core.vim [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+" the command-line testing environment did not load plugin/* in the normal
+" way.  Therefore, I do the check everywhere so I don't have to special-case
+" the command line.
+
+if !exists('g:editorconfig_core_vimscript_debug')
+    let g:editorconfig_core_vimscript_debug = 0
+endif
+" }}}1
+
+" The version we are, i.e., the latest version we support
```
Comment (xuhdev) [inline autoload/editorconfig_core.vim]: This comment looks a bit weird. If it's about the latest version of the format we support, we should only have two numbers, because the third is for software update (bug fixes). I would just simply say "the version," because that's the least confusing and people should understand what it means.

File: autoload/editorconfig_core/fnmatch.vim [hunk 4]
```diff
@@ -0,0 +1,441 @@
+" encoding is utf8 - 中文
```
Comment (xuhdev) [inline autoload/editorconfig_core/fnmatch.vim]: Is there a reason to have Chinese characters on this line?

File: autoload/editorconfig_core/ini.vim [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+        \ l:config_dirname, l:glob)
+endfunction " matches_filename
+
+" }}}1
+" === Copyright notices ================================================= {{{2
+""""EditorConfig file parser
+
+"Based on code from ConfigParser.py file distributed with Python 2.6.
+
+"Licensed under PSF License (see LICENSE.PSF file).
```
Comment (xuhdev) [inline autoload/editorconfig_core/ini.vim]: We cannot license our files under PSF license, because PSF license is specifically made for Python Software Foundation. The first sentence of the PSF license reads

> This LICENSE AGREEMENT is between the Python Software Foundation ("PSF"), and the Individual or Organization ("Licensee") ...

File: doc/editorconfig.txt [hunk 6]
```diff
@@ -8,16 +8,16 @@ Description: EditorConfig vim plugin
 License:
    Copyright (c) 2011-2015 EditorConfig Team
```
Comment (xuhdev) [inline doc/editorconfig.txt]: Can you also update the year?

File: autoload/editorconfig_core/fnmatch.vim (hunk 4 above)
Comment (cxw42) [reply autoload/editorconfig_core/fnmatch.vim]: It is my personal style, but I can remove it if you wish.  I like to include some non-ASCII characters on the first line so that programs can more easily tell it's UTF-8 text.

File: doc/editorconfig.txt (hunk 6 above)
Comment (cxw42) [reply doc/editorconfig.txt]: Sure!

**Comment by cxw42** [comment]
Question: what should the new version number be?  This plugin is currently v0.3.3.  Are we using semver? 
 Since we are removing functions, I think it should be v1.0.0 if we are using semver.  What do you think?  I will update the docs accordingly.

**Comment by xuhdev** [comment]
Yes, I think v1.0.0 is the best new version number.

**Comment by xuhdev** [comment]
Perhaps v0.99? Let's mark some instability for now.

File: autoload/editorconfig_core.vim (hunk 3 above)
Comment (cxw42) [reply autoload/editorconfig_core.vim]: I am happy to change it.  Is there a wiki page or other document defining the project's versioning conventions?  I see that v0.12.2 is the PyPI version of editorconfig-core-py.  I also see that v0.12 is the latest version of editorconfig-core-test.  Does this mean that plugins should use the core-test version as their MAJOR.MINOR, but have their own PATCH?

The [core tests](https://github.com/editorconfig/editorconfig-core-test/blob/master/cli/CMakeLists.txt#L31) require the core to return a three-component version number.  Therefore, I have to use at least `0.12.0` to pass the tests.

What is the plan for the version numbering?  Thanks for helping me understand!

**Comment by cxw42** [comment]
@xuhdev If we're actually semver, it could also be version `1.0.0-beta` to mark it as a prerelease.  OK?  If not, I'll use 0.99.

**Comment by cxw42** [comment]
Separate question - what is [`mkzip.sh`](https://github.com/editorconfig/editorconfig-vim/blob/master/mkzip.sh) used for?  I just noticed it for the first time - does it need to be updated?  It doesn't have any comments explaining its purpose.  However, it does mention Python, so I think it should be updated.

**Comment by xuhdev** [comment]
Sounds good to me.

File: autoload/editorconfig_core/fnmatch.vim (hunk 4 above)
Comment (xuhdev) [reply autoload/editorconfig_core/fnmatch.vim]: How about [IDEOGRAPHIC FULL STOP](http://www.fileformat.info/info/unicode/char/3002/index.htm)? It just looks semantically less confusing.

    encoding is utf8。

File: autoload/editorconfig_core/ini.vim (hunk 5 above)
Comment (cxw42) [reply autoload/editorconfig_core/ini.vim]: I updated the code to clarify.  We are not licensing under PSF.  However, we derived from PSF-licensed code, so have to include the PSF license text and copyright notice.

**Comment by cxw42** [comment]
@xuhdev I am logging off and will take a look at your next comments in the next few days.  Thanks for your detailed review!

File: autoload/editorconfig_core/fnmatch.vim (hunk 4 above)
Comment (cxw42) [reply autoload/editorconfig_core/fnmatch.vim]: Thanks!  I like that and have made the change.

File: autoload/editorconfig_core.vim (hunk 3 above)
Comment (cxw42) [reply autoload/editorconfig_core.vim]: @xuhdev Can we proceed with this while waiting for the general discussion in editorconfig issue 395?  vimcore currently passes all tests, including the version-related ones, and I believe this is the last open issue.  What do you think?

Comment (xuhdev) [reply autoload/editorconfig_core.vim]: Sorry, for some reason I didn't receive any notification about this. Perhaps it is the comment here that confused me ("the latest version we support" rather than "the version"). Which one do you intend it to be?

Comment (cxw42) [reply autoload/editorconfig_core.vim]: @xuhdev I just updated my comment in editorconfig 395, and I updated the comment here in 53d117a7db6ba2c3da9efcf58b0734d011aadc37.  This is intended to be the specification version.  OK?

File: autoload/editorconfig_core.vim [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+        endfor
+
+    endfor "foreach target
+
+    " Write the output file
+    call writefile(l:output, a:names.output)
+endfunction "editorconfig_core#currbuf_cli
+
+" }}}1
+" === Caching (not yet used) ============================================ {{{1
```
Comment (xuhdev) [inline autoload/editorconfig_core.vim]: What do you mean by "not yet used"? Can we simply remove these code if they half-baken? We can make another PR to track the enabling of caching.

File: autoload/editorconfig_core.vim (hunk 3 above)
Comment (cxw42) [reply autoload/editorconfig_core.vim]: Marked as resolved per your comment in https://github.com/editorconfig/editorconfig-vim/pull/119#pullrequestreview-238690349

File: autoload/editorconfig_core.vim (hunk 7 above)
Comment (cxw42) [reply autoload/editorconfig_core.vim]: Removed in https://github.com/editorconfig/editorconfig-vim/pull/119/commits/6a229d07c6e72019c80cd165c7511c7672bafb8e

**Comment by cxw42** [comment]
@xuhdev https://github.com/editorconfig/editorconfig-vim/pull/119#pullrequestreview-238690349

> Regarding to your new changes, I've made another comment. You also seem to have changed EOL of a few files. Just wanna make sure you intended it. 

Would you please let me know which files you are looking at?  I just looked through `git diff upstream/master` and didn't see any EOL changes.  All files currently have Unix (LF) line endings, which I think is reasonable.  (Except for the `crlf.in` file from editorconfig-core-test, which needs to have CRLF to function! :) )

There are a number of renames.  For example, I moved `tests/` to `tests/plugin` so that I could create `tests/core`.  Is that perhaps what you are looking at?

**Comment by xuhdev** [comment]
Perhaps I'm getting confused. Could you tell me, for example, what has changed in `tests/core/ecvimlib.ps1` since then? I've seen a whole file change [here](https://github.com/editorconfig/editorconfig-vim/commit/53d117a7db6ba2c3da9efcf58b0734d011aadc37#diff-71263ae416d5a3f441fb25172968cb4e).

**Comment by cxw42** [comment]
Ah!  Sure.  Compared to `master`, that is a new file.  Everything in`tests/core` is copied in from `editorconfig-core-vimscript` per your earlier request to combine the repos.  But I see that GitHub, at least on mobile, does not make that clear at all!

**Comment by cxw42** [comment]
@xuhdev I cleaned up and squashed some.  The CRLF files are now consistent throughout the commits in which they occur, and the `.editorconfig` file specifies those line ends appropriately.  The previous commits are still available in [cxw42/vimcore-old](https://github.com/cxw42/editorconfig-vim/tree/vimcore-old) if you want to compare.

**Comment by xuhdev** [comment]
Thank you for your hard work!

**Comment by cxw42** [comment]
@xuhdev Happy to help!  

When you have a chance, would you please set up Appveyor for this repo under editorconfig and update README.md?  It currently points to my Appveyor account.  No hurry, but I think it would be a good idea to continue running those tests.

**Comment by xuhdev** [comment]
Done!

**Comment by xuhdev** [comment]
Sorry for the late discussion on this issue, but I realized that we should have retained the ability to use an external core library as an option (the one that calls the external editorconfig executable, not the Python builtin). Could you restore the availability of that option, if this is not too difficult? We have done something similar to Emacs before: https://github.com/editorconfig/editorconfig-emacs/pull/46#issuecomment-160042488 I'm OK with defaulting to the internal one, but an option to use an external one might still be important.
